### PR TITLE
refactor: use `NonZeroU*` where appropriate

### PR DESCRIPTION
**Pull request by mlegner**
In particular, replace `u16` and `u32` for symbol counts, symbol sizes, and number of shards by `NonZeroU16`.

TODO:
- [x] If #203 is merged before, refactor that code to use the same types.

### CODE REVIEW

File: crates/walrus-core/src/encoding/blob_encoding.rs [hunk 1]
```diff
@@ -237,11 +242,10 @@ impl<'a> ExpandedMessageMatrix<'a> {
...
     /// Fills the systematic part of the matrix using `self.rows`.
     fn fill_systematic_with_rows(&mut self) {
-        for (destination_row, row) in self
-            .matrix
-            .iter_mut()
-            .zip(self.blob.chunks(self.n_columns * self.symbol_size as usize))
-        {
+        for (destination_row, row) in self.matrix.iter_mut().zip(
+            self.blob
+                .chunks(self.n_columns * self.symbol_size.get() as usize),
```
Comment (giac-mysten) [inline crates/walrus-core/src/encoding/blob_encoding.rs:247]: Remark: strange change in formatting but ok 😅

File: crates/walrus-core/src/encoding/blob_encoding.rs [hunk 2]
```diff
@@ -172,12 +181,12 @@ impl<'a> BlobEncoder<'a> {
     }
 
     fn empty_slivers<T: EncodingAxis>(&self) -> Vec<Sliver<T>> {
-        (0..self.config.n_shards)
+        (0..self.config.n_shards().get())
```
Comment (giac-mysten) [inline crates/walrus-core/src/encoding/blob_encoding.rs:184]: Remark: technically here you don't need the function?

File: crates/walrus-service/src/client.rs [hunk 3]
```diff
@@ -185,7 +185,7 @@ impl Client {
             self.committee.epoch,
             &self.client,
             node,
-            self.committee.total_weight,
+            self.committee.total_weight.into(),
```
Comment (giac-mysten) [inline crates/walrus-service/src/client.rs]: Nit: Could you change `NodeCommunication` `total_weight` to `NonZeroU16` as well? or I can do that in a next pr

File: crates/walrus-core/src/metadata.rs [hunk 4]
```diff
@@ -157,16 +154,13 @@ impl VerifiedBlobMetadataWithId {
 impl UnverifiedBlobMetadataWithId {
     /// Consumes the metadata and attempts to verify it the relationship between the contained
     /// metadata and blob ID. On success, returns a [`VerifiedBlobMetadataWithId`].
-    pub fn verify(
-        self,
-        n_shards: NonZeroUsize,
-    ) -> Result<VerifiedBlobMetadataWithId, VerificationError> {
+    pub fn verify(self, n_shards: usize) -> Result<VerifiedBlobMetadataWithId, VerificationError> {
```
Comment (giac-mysten) [inline crates/walrus-core/src/metadata.rs]: Nit: How come `usize` here? not that it matters, my pr changes this

Comment (mlegner) [reply crates/walrus-core/src/metadata.rs]: Probably no good reason, might have been an oversight. 😅

File: crates/walrus-core/src/encoding/blob_encoding.rs (hunk 1 above)
Comment (mlegner) [reply crates/walrus-core/src/encoding/blob_encoding.rs:247]: 🤷

File: crates/walrus-core/src/encoding/basic_encoding.rs [hunk 5]
```diff
@@ -120,17 +117,20 @@ impl Encoder {
         self.raptorq_encoder
             .source_packets()
             .into_iter()
-            .chain(
-                self.raptorq_encoder
-                    .repair_packets(0, self.n_shards - self.n_source_symbols as u32),
-            )
+            .chain(self.raptorq_encoder.repair_packets(
+                0,
+                (self.n_shards.get() - self.n_source_symbols.get()).into(),
```
Comment (jpcsmith) [inline crates/walrus-core/src/encoding/basic_encoding.rs:128]: Question: Does this class have an invariant that n_shards > n_source_symbols?

File: crates/walrus-core/src/lib.rs [hunk 6]
```diff
@@ -237,10 +237,13 @@ impl Sliver {
...
     pub fn has_correct_length(&self, config: &EncodingConfig, blob_size: usize) -> bool {
-        Some(self.len()) == self.expected_length(config, blob_size)
+        Some(self.len())
+            == self
+                .expected_length(config, blob_size)
+                .map(NonZeroUsize::get)
     }
 
-    fn expected_length(&self, config: &EncodingConfig, blob_size: usize) -> Option<usize> {
+    fn expected_length(&self, config: &EncodingConfig, blob_size: usize) -> Option<NonZeroUsize> {
```
Comment (jpcsmith) [inline crates/walrus-core/src/lib.rs]: Remark: The NonZeroUsize probably wasn't necessary here.

File: crates/walrus-core/benches/constants/mod.rs [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-pub const N_SHARDS: u32 = 1000;
+pub const N_SHARDS: u16 = 1000;
```
Comment (jpcsmith) [inline crates/walrus-core/benches/constants/mod.rs:4]: Remark: This could also be `= unsafe { NonZeroU16::new_unchecked(1000) }` if it helps simplifying usage elsewhere.

File: crates/walrus-service/src/node.rs [hunk 8]
```diff
@@ -165,18 +165,21 @@ pub struct StorageNode<T> {
 impl StorageNode<SuiReadClient> {
     /// Create a new storage node with the provided configuration.
     pub async fn new(
-        config: &StorageNodeConfig,
+        node_config: &StorageNodeConfig,
```
Comment (jpcsmith) [inline crates/walrus-service/src/node.rs:168]: Question: Why the change of name?

File: crates/walrus-service/src/mapping.rs [hunk 9]
```diff
@@ -148,8 +149,12 @@ mod tests {
         let num = num as u16;
         (0..num)
             .map(|n| SliverPair {
-                primary: Sliver::new_empty(0, 1, SliverPairIndex::new(n)),
-                secondary: Sliver::new_empty(0, 1, SliverPairIndex::new(num - n - 1)),
+                primary: Sliver::new_empty(0, 1.try_into().unwrap(), SliverPairIndex::new(n)),
+                secondary: Sliver::new_empty(
+                    0,
+                    1.try_into().unwrap(),
+                    SliverPairIndex::new(num - n - 1),
```
Comment (jpcsmith) [inline crates/walrus-service/src/mapping.rs]: Question: Is there a function on the Sliver*Index doing this conversion already? If so, it would be best not to duplicate the logic it here.

File: crates/walrus-core/src/encoding/utils.rs [hunk 10]
```diff
@@ -1,17 +1,20 @@
...
+
 use raptorq::{EncodingPacket, ObjectTransmissionInformation};
 
 /// Creates a new [`ObjectTransmissionInformation`] for the given `symbol_size`.
 ///
 /// # Panics
 ///
 /// Panics if the `symbol_size` is larger than [`MAX_SYMBOL_SIZE`][super::MAX_SYMBOL_SIZE].
-pub fn get_transmission_info(symbol_size: usize) -> ObjectTransmissionInformation {
+pub fn get_transmission_info(symbol_size: NonZeroUsize) -> ObjectTransmissionInformation {
```
Comment (jpcsmith) [inline crates/walrus-core/src/encoding/utils.rs]: Question: Why usize here?

File: crates/walrus-core/src/encoding/slivers.rs [hunk 11]
```diff
@@ -49,8 +48,8 @@ impl<T: EncodingAxis> Sliver<T> {
     /// # Panics
     ///
     /// Panics if the slice does not contain complete symbols, i.e., if
-    /// `slice.len() % symbol_size != 0` or if `symbol_size == 0`.
-    pub fn new<U: Into<Vec<u8>>>(data: U, symbol_size: u16, index: SliverIndex) -> Self {
+    /// `slice.len() % symbol_size != 0`.
```
Comment (jpcsmith) [inline crates/walrus-core/src/encoding/slivers.rs]:
```suggestion
    /// `data.len() % symbol_size != 0`.
```

File: crates/walrus-core/src/encoding/slivers.rs [hunk 12]
```diff
@@ -227,7 +222,12 @@ impl<T: EncodingAxis> Sliver<T> {
         if symbol_size == 0 || inner_iter.any(|s| s.data.len() != symbol_size) {
             return Err(RecoveryError::InvalidSymbolSizes);
         }
-        let symbol_size: u16 = symbol_size.try_into().map_err(|_| DataTooLargeError)?;
+        let symbol_size = NonZeroU16::new(
+            symbol_size
+                .try_into()
+                .map_err(|_| EncodeError::DataTooLarge)?,
+        )
+        .expect("already checked size above");
```
Comment (jpcsmith) [inline crates/walrus-core/src/encoding/slivers.rs]: Remark: Not sure if this is cleaner:
```suggestion
        let symbol_size = symbol_size
            .try_into()
            .ok()
            .and_then(NonZeroU16::new)
            .ok_or(EncodeError::DataTooLarge)?;
```

File: crates/walrus-core/src/encoding/blob_encoding.rs (hunk 2 above)
Comment (mlegner) [reply crates/walrus-core/src/encoding/blob_encoding.rs:184]: Unfortunately, the range seems to expect an integer and doesn't work without the `get()`.

File: crates/walrus-service/src/client.rs (hunk 3 above)
Comment (mlegner) [reply crates/walrus-service/src/client.rs]: I actually removed it, as we now have the encoding config there.

File: crates/walrus-core/src/encoding/basic_encoding.rs (hunk 5 above)
Comment (mlegner) [reply crates/walrus-core/src/encoding/basic_encoding.rs:128]: Thanks, I added that.

File: crates/walrus-service/src/node.rs (hunk 8 above)
Comment (mlegner) [reply crates/walrus-service/src/node.rs:168]: To differentiate it from the `encoding_config`.

File: crates/walrus-service/src/mapping.rs (hunk 9 above)
Comment (mlegner) [reply crates/walrus-service/src/mapping.rs]: Good point, I changed this to reuse the `SliverPair::new_empty` function.

File: crates/walrus-core/src/encoding/utils.rs (hunk 10 above)
Comment (mlegner) [reply crates/walrus-core/src/encoding/utils.rs]: No good reason. 😇

File: crates/walrus-core/benches/constants/mod.rs (hunk 7 above)
Comment (mlegner) [reply crates/walrus-core/benches/constants/mod.rs:4]: Thanks for the hint, but I think it's currently not needed.
